java-checks/src/main/java/org/sonar/java/checks/UselessPackageInfoCheck.java

```diff
@@ -20,23 +20,57 @@
 package org.sonar.java.checks;
 
 import java.io.File;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
 import org.sonar.check.Rule;
+import org.sonar.java.EndOfAnalysisCheck;
+import org.sonar.java.checks.helpers.ExpressionsHelper;
 import org.sonar.plugins.java.api.JavaFileScanner;
 import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.PackageDeclarationTree;
 
 @Rule(key = "S4032")
-public class UselessPackageInfoCheck implements JavaFileScanner {
+public class UselessPackageInfoCheck implements JavaFileScanner, EndOfAnalysisCheck {
+
+  private final Map<String, JavaFileScannerContext> unneededPackageInfoFiles = new HashMap<>();
+  private final Set<String> knownPackagesWithOtherFiles = new HashSet<>();
 
   @Override
   public void scanFile(JavaFileScannerContext context) {
-    File file = context.getInputFile().file();
-    if ("package-info.java".equals(file.getName()) && isOnlyFileFromPackage(file)) {
-      context.addIssueOnFile(this, "Remove this package.");
+    PackageDeclarationTree packageDeclaration = context.getTree().packageDeclaration();
+    String packageName = packageDeclaration == null ? null : ExpressionsHelper.concatenate(packageDeclaration.packageName());
+
+    // default package or already processed package
+    if (packageName == null || knownPackagesWithOtherFiles.contains(packageName)) {
+      return;
+    }
+
+    File packageDirectory = context.getInputFile().file().getParentFile();
+    File packageInfoFile = new File(packageDirectory, "package-info.java");
+    boolean hasOtherFiles = !isOnlyFileFromPackage(packageDirectory, packageInfoFile);
+
+    if (hasOtherFiles) {
+      knownPackagesWithOtherFiles.add(packageName);
+    } else if (packageInfoFile.isFile()) {
+      unneededPackageInfoFiles.put(packageName, context);
+    }
+  }
+
+  @Override
+  public void endOfAnalysis() {
+    unneededPackageInfoFiles.keySet().removeAll(knownPackagesWithOtherFiles);
```

Comment:
The behavior of `scanFile` seems to be that it adds a package to one or the other but not both. Is the explicit removal of valid packages from the the set of unneeded packages necessary?

Reply:
Yes, it is needed to avoid FP in case the analysis scan the same package multiple times (from different source folder). In such a case, the package may be present in both collections, but we do not want to raise an issue.